### PR TITLE
[Quantization] Dequantize(rescale) -> Dequantize()

### PR DESCRIPTION
**Pull request by tlepley-cadence**
*Description*: Add optim: Dequantize(rescale) -> Dequantize()
We've got this case in one of our models with our backend.

*Testing*: Added test: mergeRescaleIntoDequantize

*Documentation*: N/A

### CODE REVIEW

File: tests/unittests/graphOptzTest.cpp [hunk 1]
```diff
@@ -1554,6 +1554,29 @@ TEST_F(GraphOptz, foldQuantizeIntoVarMultipleUsages) {
   }
 }
```
Comment (jfix71) [inline tests/unittests/graphOptzTest.cpp]: Can you also add a doxygen comment for the test?

File: tests/unittests/graphOptzTest.cpp [hunk 2]
```diff
@@ -1554,6 +1554,29 @@ TEST_F(GraphOptz, foldQuantizeIntoVarMultipleUsages) {
...
 
+TEST_F(GraphOptz, mergeRescaleIntoDequantize) {
+  // Check that we are combining quantization-dequantization pairs.
+  auto *input = mod_.createPlaceholder(ElemKind::Int8QTy, {4, 10}, 0.5, 11,
+                                       "input", true);
+  auto qType = mod_.uniqueType(ElemKind::Int8QTy, {4, 10}, 0.03f, 5);
```
Comment (rdzhabarov) [inline tests/unittests/graphOptzTest.cpp]: nit: auto *qType

File: tests/unittests/graphOptzTest.cpp [hunk 3]
```diff
@@ -1554,6 +1554,29 @@ TEST_F(GraphOptz, foldQuantizeIntoVarMultipleUsages) {
...
+  auto *R = F_->createRescaleQuantized("rescale", input, qType);
+  auto *D = F_->createDequantize("dequantize", R);
+  F_->createSave("ret", D);
+
+  EXPECT_EQ(F_->getNodes().size(), 3);
+  ::glow::optimize(F_, CompilationMode::Infer);
+  EXPECT_EQ(F_->getNodes().size(), 2);
+
+  // Check that that we get the expected set of nodes
```
Comment (rdzhabarov) [inline tests/unittests/graphOptzTest.cpp]: grammar

File: tests/unittests/graphOptzTest.cpp (hunk 1 above)
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Done.

File: tests/unittests/graphOptzTest.cpp (hunk 2 above)
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Done. 
`uniqueType` returns a `TypeRef` object that at first sight doesn't look like a pointer. It turns out that `TypeRef` is a pointer to `Type`, but calling it `TypePtr` would certainly be more explicit (a 'ref' is usually not a pointer).

File: tests/unittests/graphOptzTest.cpp (hunk 3 above)
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Fixed

File: tests/unittests/graphOptzTest.cpp [hunk 4]
```diff
@@ -1554,6 +1554,29 @@ TEST_F(GraphOptz, foldQuantizeIntoVarMultipleUsages) {
...
+  bool hasDequantize = false;
+  bool hasSave = false;
+  for (auto &node : F_->getNodes()) {
+    hasDequantize |= (node.getKind() == Kinded::Kind::DequantizeNodeKind);
+    hasSave |= (node.getKind() == Kinded::Kind::SaveNodeKind);
```
[the remark this answers is not in the document]
Comment (tlepley-cadence) [reply tests/unittests/graphOptzTest.cpp]: Done. Note that it was already better than the tests around that usually just test the number of nodes ;)

**Comment by nadavrot** [comment]
Nice!

**Pull request merged**